File: update_links.py
import os
import re
from bs4 import BeautifulSoup

def remove_html_extension_from_links(html_content):
    # Parse the HTML
    soup = BeautifulSoup(html_content, 'html.parser')
    
    # Find all <a> tags with href attributes
    for a_tag in soup.find_all('a', href=True):
        href = a_tag['href']
        
        # Skip external links, anchors, and mailto: links
        if href.startswith(('http', 'mailto:', 'tel:', '#', 'javascript:')) or '//' in href:
            continue
            
        # Remove .html from internal links
        if href.endswith('.html'):
            a_tag['href'] = href[:-5]  # Remove .html
        elif href.endswith('/index'):
            a_tag['href'] = href[:-5]  # Remove /index
            
    return str(soup)

def process_html_files(directory):
    # Get all HTML files in the directory
    for root, _, files in os.walk(directory):
        for file in files:
            if file.endswith('.html'):
                file_path = os.path.join(root, file)
                print(f"Processing: {file_path}")
                
                # Read the file
                with open(file_path, 'r', encoding='utf-8') as f:
                    content = f.read()
                
                # Update links
                updated_content = remove_html_extension_from_links(content)
                
                # Write the updated content back to the file
                with open(file_path, 'w', encoding='utf-8') as f:
                    f.write(updated_content)

if __name__ == "__main__":
    # Get the directory where this script is located
    script_dir = os.path.dirname(os.path.abspath(__file__))
    
    # Process all HTML files in the script's directory
    process_html_files(script_dir)
    
    print("\nLink update complete!")
    print("Don't forget to test all links after updating.")

File: updateLinks.js
const fs = require('fs');
const path = require('path');

// Function to process a single HTML file
function processFile(filePath) {
  try {
    console.log(`Processing: ${filePath}`);
    
    // Read the file
    let content = fs.readFileSync(filePath, 'utf8');
    
    // Replace .html in href attributes (handles both single and double quotes)
    content = content.replace(
      /href=(["'])([^"']+)\.html(["'])/g,
      'href=$1$2$3'
    );
    
    // Also handle links that end with /index
    content = content.replace(
      /href=(["'])([^"']+)\/index(["'])/g,
      'href=$1$2$3'
    );
    
    // Write the updated content back to the file
    fs.writeFileSync(filePath, content, 'utf8');
    return true;
  } catch (error) {
    console.error(`Error processing ${filePath}:`, error);
    return false;
  }
}

// Function to find all HTML files in a directory
function processDirectory(directory) {
  const files = fs.readdirSync(directory);
  let processedCount = 0;
  let errorCount = 0;
  
  files.forEach(file => {
    const fullPath = path.join(directory, file);
    const stat = fs.statSync(fullPath);
    
    if (stat.isDirectory()) {
      // Skip node_modules and other common directories
      if (!['node_modules', '.git', '.github', '.vscode'].includes(file)) {
        const result = processDirectory(fullPath);
        processedCount += result.processedCount;
        errorCount += result.errorCount;
      }
    } else if (file.endsWith('.html')) {
      const success = processFile(fullPath);
      if (success) {
        processedCount++;
      } else {
        errorCount++;
      }
    }
  });
  
  return { processedCount, errorCount };
}

// Start processing from the current directory
console.log('Starting to update HTML links...');
const { processedCount, errorCount } = processDirectory(__dirname);

console.log('\n--- Update Complete ---');
console.log(`Processed: ${processedCount} files`);
if (errorCount > 0) {
  console.warn(`Encountered errors with ${errorCount} files`);
} else {
  console.log('All files processed successfully!');
}
console.log('\nPlease test all links to ensure they work correctly.');
